image.py

```diff
@@ -0,0 +1,355 @@
+import io
+from shlex import split
+from subprocess import CalledProcessError, run
+from sys import stdin
+from typing import Union
+
+from exceptions import CommandNotFoundOnImageError
+
+
+class Image:
+    """A class which holds a reference to a given docker image and provides
+    an interface by which to interact with that image.
+
+    -   Capable of building docker images from dockerfiles or dockerfile-
+        formatted strings via the build classmethods.
+    -   Capable of acquiring information about the given image (see property
+        methods below).
+    -   Capable of running commands on containers built from the image using
+        the run method.
+    -   All docker build and run methods on the class can recive a file-like
+        object, (io.TextIOBase or some subclass of it,) and they will write the
+        outputs from docker into the file to enable logging.
+
+    NOTE: This class does not work correctly if the image built does not have
+        bash installed.
+    """
+
+    def __init__(self, name_or_id: str):
+        """Initialize this Image. Connects the class with its ID.
+
+        Parameters
+        ----------
+        name_or_id : str
+            A name or ID by which to find this image using docker inspect.
+        """
+        self._id = get_image_id(name_or_id)
+
+    @classmethod
+    def build_from_dockerfile(
+        cls,
+        context: str,
+        tag: str,
+        output_file: Union[io.TextIOBase, None] = None,
+        dockerfile_loc: str = ""
+    ):
```

Comment:
Regarding this, I found that mypy didn't like this method and threw up a lot of complaints - I used `from typing_extensions import Self` and returned this as a Self variable, which the internet tells me is supported on Python 3.7+?

Reply:
Ah sorry, I think the problem is that mypy isn't smart enough to infer the type of `cls`, so we have to manually annotate it.

Here's a minimal example that works for me:

```python
# self_type.py
from typing import Type, TypeVar

class Foo:
    Self = TypeVar("Self", bound="Foo")

    @classmethod
    def factory(cls: Type[Self]) -> Self:  # <-- Note the annotation of cls: Type[Self]
        return cls()

if __name__ == "__main__":
    foo = Foo.factory()
```

```console
$ python --version
Python 3.8.16
$ mypy --version
mypy 1.0.1 (compiled: yes)
$ mypy self_type.py
Success: no issues found in 1 source file
```

Try adding a similar annotation for `cls` in your parameter list and see if it works then.

> I used `from typing_extensions import Self` and returned this as a Self variable, which the internet tells me is supported on Python 3.7+?

Presumably this would require adding a new dependency on the `typing_extensions` package, so I'd like to avoid this.